wasm/jit/engine.go

```diff
@@ -0,0 +1,329 @@
+	// Wasm stack pointer on .stack field which is accessed by currentBaseStackPointer+currentStackPointer
+	currentStackPointer     uint64
+	currentBaseStackPointer uint64
+	// Where we store the status code of JIT execution.
+	jitCallStatusCode jitStatusCodes
+	// Set when statusCode == jitStatusCall{Function,BuiltInFunction,HostFunction}
+	// Indicating the function call index.
+	functionCallIndex int64
+	// Set when statusCode == jitStatusCall{Function,BuiltInFunction,HostFunction}
+	// We use this value to continue the current function
+	// after calling the target function exits.
+	// Instructions after [base+continuationAddressOffset] must start with
+	// restoring reserved registeres.
+	continuationAddressOffset uintptr
+	// Function call frames in linked list
+	callFrameStack *callFrame
+	// Store the compiled functions and indexes.
+	compiledWasmFunctions     []*compiledWasmFunction
+	compiledWasmFunctionIndex map[*wasm.FunctionInstance]int64
+	// Store the host functions and indexes.
+	hostFunctions     []func(ctx *wasm.HostFunctionCallContext)
+	hostFunctionIndex map[*wasm.FunctionInstance]int64
+}
+
+func (e *engine) Call(f *wasm.FunctionInstance, args ...uint64) (returns []uint64, err error) {
+	for _, arg := range args {
+		e.push(arg)
+	}
+	if index, ok := e.hostFunctionIndex[f]; ok {
+		e.hostFunctions[index](&wasm.HostFunctionCallContext{Memory: f.ModuleInstance.Memory})
+	} else if index, ok := e.compiledWasmFunctionIndex[f]; ok {
+		f := e.compiledWasmFunctions[index]
+		e.exec(f)
+	} else {
+		err = fmt.Errorf("invalid function")
+		return
+	}
+	returns = make([]uint64, len(f.Signature.ReturnTypes))
+	for i := range returns {
+		returns[len(returns)-1-i] = e.pop()
+	}
+	return
+}
+
+// PreCompile implements wasm.Engine for engine.
+// Here we assign unique ids to all the function instances,
+// so we can reference it when we compile each function instance.
+func (e *engine) PreCompile(fs []*wasm.FunctionInstance) error {
+	var newUniqueHostFunctions, newUniqueWasmFunctions int
+	for _, f := range fs {
+		if f.HostFunction != nil {
+			if _, ok := e.hostFunctionIndex[f]; ok {
+				continue
+			}
+			id := int64(len(e.hostFunctionIndex))
+			e.hostFunctionIndex[f] = id
+			newUniqueHostFunctions++
+		} else {
+			if _, ok := e.compiledWasmFunctionIndex[f]; ok {
+				continue
+			}
+			id := int64(len(e.compiledWasmFunctionIndex))
+			e.compiledWasmFunctionIndex[f] = id
+			newUniqueWasmFunctions++
+		}
+	}
+	e.hostFunctions = append(
+		e.hostFunctions,
+		make([]func(ctx *wasm.HostFunctionCallContext), newUniqueHostFunctions)...,
+	)
+	e.compiledWasmFunctions = append(
+		e.compiledWasmFunctions,
+		make([]*compiledWasmFunction, newUniqueWasmFunctions)...,
+	)
+	return nil
+}
+
+func (e *engine) Compile(f *wasm.FunctionInstance) error {
+	if f.HostFunction != nil {
+		id := e.hostFunctionIndex[f]
+		if e.hostFunctions[id] != nil {
+			// Already compiled.
+			return nil
+		}
+		hf := func(ctx *wasm.HostFunctionCallContext) {
```

Comment:
coud this be extracted into a type, or is it not worth it?

Reply:
https://github.com/tetratelabs/wazero/pull/60/commits/eae931493c14b9c968a7876b7ee8bb6c84ab240c done